libs/ec/inc/public/tfc/devices/base.hpp

```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <boost/asio.hpp>
+
+#include "tfc/ecx.hpp"
+
+namespace tfc::ec::devices {
+class base {
+public:
+  explicit base(boost::asio::io_context& ctx) : ctx_(ctx) {}
+  static constexpr auto vendor_id = 0x0;
+  static constexpr auto product_code = 0x0;
+  virtual ~base() = default;
+  // Default behaviour no data processing
+  virtual void process_data(uint8_t*, uint8_t*) noexcept {};
+  // Default behaviour, no setup
+  virtual auto setup(ecx_contextt*, uint16_t) -> int { return 1; };
+
+protected:
+  boost::asio::io_context& ctx_;
```

Comment:
do we need this reference?, I don't think it is nice to keep references in long living classes, but this should probably be safe.

Reply:
The only place these classes reference the ctx is to construct signals and slots in the setup method. I want to change the constructor to accept the slave_index at construction.

Then i can create all slots/signals in the constructor and can remove the reference.

Reply:
I have implemented this change. This also has the added benefit that future device interfaces are less likely to construct signals/slots in the setup method.